### PR TITLE
refactor-improve-indexers

### PR DESCRIPTION
**Pull request by numb3r3**
**Changes**:
    
- use `limit`, deprecate `top_k`
- use `add_with_ids` and `remove_ids` to improve **update** and **delete** implement
- support `ef_query` and `ef_construct` parameter for HNSW index
- add `/status` and `/clear` endpoints into FaissSearcher
- use `index_traversal_paths` and `search_traversal_paths` as parameters
- refactor online `/train` endpoint
    - use PSQL to store trained index
    - fetch the stored index from PSQL at initialization step
   

TODO:
- [x] integrate test for online & offline training. refer to https://github.com/jina-ai/executors/blob/20a86270c2705b83ebdb2b14cf3809ed1019ee06/tests/integration/faiss_psql/test_train_psql.py

### CODE REVIEW

File: jinahub/indexers/storage/PostgreSQLStorage/postgreshandler.py [hunk 1]
```diff
@@ -378,69 +430,103 @@ def get_snapshot(self, shards_to_get: List[int]):
...
+            self.logger.error(f'Error importing snapshot: {error}')
+            self.connection.rollback()
+        self.connection.commit()
+
+    def get_document_iterator(
+        self,
+        limit: int = 0,
+        return_embedding: bool = False,
+        check_embedding: bool = False,
+    ) -> Generator[Document, None, None]:
```
Comment (cristianmtr) [inline jinahub/indexers/storage/PostgreSQLStorage/postgreshandler.py:444]: docstring here too

File: jinahub/indexers/storage/PostgreSQLStorage/tests/test_postgres_dbms.py [hunk 2]
```diff
@@ -184,6 +184,51 @@ def test_return_embeddings(docker_compose):
...
+                chunks=0,
+                same_content=False,
+            )
+        )
+    )
+
+    indexer.add(docs)
+    assert len(list(indexer.get_document_iterator())) == NR
+
+    indexer.clear()
```
Comment (cristianmtr) [inline jinahub/indexers/storage/PostgreSQLStorage/tests/test_postgres_dbms.py:228]: Seeing this test, I wonder if right name here should just be `delete_all`

File: jinahub/indexers/storage/PostgreSQLStorage/tests/test_postgres_dbms.py [hunk 3]
```diff
@@ -277,15 +322,12 @@ def _assert_snapshot_shard_distribution(func, nr_shards, total_docs_expected):
...
 
     # we use total_shards=1 in order to guarantee getting all the data in the delta
-    deltas = postgres_indexer._get_delta(
+    deltas = postgres_indexer.get_delta_updates(
         shard_id=0, total_shards=1, timestamp=timestamp
     )
     deltas = list(deltas)
     np.testing.assert_equal(len(deltas), NR_DOCS_DELTA + NR_DOCS_DELTA_DELETED)
 
-    # TODO delta is then passed to the indexer
```
Comment (cristianmtr) [inline jinahub/indexers/storage/PostgreSQLStorage/tests/test_postgres_dbms.py:286]: If you are deleting this, it means Faiss can now do soft-deletion?

Comment (cristianmtr) [reply jinahub/indexers/storage/PostgreSQLStorage/tests/test_postgres_dbms.py:286]: If so, expand the test pleae

File: tests/integration/psql_import/test_import_psql.py [hunk 4]
```diff
@@ -389,15 +390,17 @@ def test_psql_sync_delta(
...
+            parameters={'limit': len(docs_search)},
             return_results=True,
         )
         # then we assert the contents include the latest
         # perfect matches
         assert len(results[0].docs) > 0
-        for d in results[0].docs:
-            np.testing.assert_almost_equal(d.matches[0].embedding, d.embedding)
+
+        # TODO: fix issue on jina core
```
Comment (cristianmtr) [inline tests/integration/psql_import/test_import_psql.py:400]: Link to issue?

File: tests/integration/psql_import/test_import_psql.py [hunk 5]
```diff
@@ -389,15 +390,17 @@ def test_psql_sync_delta(
...
+        # for d in results[0].docs:
+        #     np.testing.assert_almost_equal(d.matches[0].embedding, d.embedding)
 
     idx = PostgreSQLStorage()
-    # size stays the same because it was only soft delete
-    assert idx.size == nr_docs
+
+    assert idx.size == nr_docs - nr_docs_to_delete
```
Comment (cristianmtr) [inline tests/integration/psql_import/test_import_psql.py:406]: I think the size should stay the same if it's soft delete. The docs are still there, right?

**Review by cristianmtr — CHANGES_REQUESTED** [review]

**Comment by cristianmtr** [comment]
We also need more thorough testing of these new features:

- [ ] integration test with separate Flows (one for Index + Search; one for Training)
  - [ ] with different Faiss index keys 
- [ ] training / storing / retrieving the model

File: jinahub/indexers/storage/PostgreSQLStorage/postgres_indexer.py [hunk 6]
```diff
@@ -98,64 +97,73 @@ def snapshot_size(self):
...
-        )
+
+        traversal_paths = parameters.get('traversal_paths', self.index_traversal_paths)
         with self.handler as postgres_handler:
             postgres_handler.update(docs.traverse_flat(traversal_paths))
 
-    @requests(on='/delete')
-    def cleanup(self, **kwargs):
+    @requests(on='/prune')
+    def prune(self, **kwargs):
```
[the remark this answers is not in the document]
Comment (numb3r3) [reply jinahub/indexers/storage/PostgreSQLStorage/postgres_indexer.py:127]: As in other indexers, `/clear` means to hard-delete all docs. `/prune` is exposed to delete the docs with soft-delete  mark, and hence to save disk space. `/clear` and `/prune` have different purpose.

File: jinahub/indexers/storage/PostgreSQLStorage/postgres_indexer.py [hunk 7]
```diff
@@ -231,11 +241,34 @@ def get_snapshot(self, shard_id: int, total_shards: int):
             )
 
             with self.handler as postgres_handler:
-                return postgres_handler.get_snapshot(shards_to_get)
+                return postgres_handler.get_snapshot(
```
[the remark this answers is not in the document]
Comment (numb3r3) [reply jinahub/indexers/storage/PostgreSQLStorage/postgres_indexer.py:244]: From my understanding, `snapshot` means a clone of all data.

File: jinahub/indexers/storage/PostgreSQLStorage/postgres_indexer.py [hunk 8]
```diff
@@ -258,26 +291,26 @@ def _vshards_to_get(shard_id, total_shards, virtual_shards):
             ]
         return [str(shard_id) for shard_id in shards_to_get]
 
-    def _get_delta(self, shard_id, total_shards, timestamp):
+    def get_delta_updates(
+        self, shard_id, total_shards, timestamp, filter_deleted: bool = False
```
[the remark this answers is not in the document]
Comment (numb3r3) [reply jinahub/indexers/storage/PostgreSQLStorage/postgres_indexer.py:295]: this is used for syncing faiss with deletions.

File: jinahub/indexers/storage/PostgreSQLStorage/postgreshandler.py [hunk 9]
```diff
@@ -90,7 +90,7 @@ def _init_table(self):
         Create table if needed with id, vecs and metas.
         """
         with self:
-            self._create_schema_version()
+            self._create_meta_table()
```
[the remark this answers is not in the document]
Comment (numb3r3) [reply jinahub/indexers/storage/PostgreSQLStorage/postgreshandler.py:93]: This table also records the `SCHEMA_VERSION`, that's the reason to name it as `meta_table`

File: jinahub/indexers/storage/PostgreSQLStorage/postgreshandler.py [hunk 10]
```diff
@@ -126,9 +128,14 @@ def _create_table(self):
                 embedding BYTEA,
                 doc BYTEA,
                 shard int,
-                last_updated timestamp with time zone default current_timestamp
+                created_at timestamp with time zone default current_timestamp,
```
[the remark this answers is not in the document]
Comment (numb3r3) [reply jinahub/indexers/storage/PostgreSQLStorage/postgreshandler.py:131]: yes, `created_at` is not used now. But I want to reserve this column for potential future updates.

File: jinahub/indexers/storage/PostgreSQLStorage/postgreshandler.py [hunk 11]
```diff
@@ -126,9 +128,14 @@ def _create_table(self):
...
+                updated_at timestamp with time zone default current_timestamp,
+                is_deleted BOOL DEFAULT FALSE
             );
-            INSERT INTO {SCHEMA_VERSIONS_TABLE_NAME} VALUES (%s, %s);''',
+            CREATE INDEX IF NOT EXISTS idx_shard ON {self.table}(shard);
```
[the remark this answers is not in the document]
Comment (numb3r3) [reply jinahub/indexers/storage/PostgreSQLStorage/postgreshandler.py:135]: The purpose of indexing on these columns aims to speed up syncing and get deltas. I will offer the comparison result on this change later.

File: jinahub/indexers/storage/PostgreSQLStorage/postgreshandler.py [hunk 12]
```diff
@@ -220,7 +227,8 @@ def update(self, docs: DocumentArray, *args, **kwargs):
             f'UPDATE {self.table}\
              SET embedding = %s,\
              doc = %s,\
-             last_updated = current_timestamp \
+             is_deleted = false, \
```
[the remark this answers is not in the document]
Comment (numb3r3) [reply jinahub/indexers/storage/PostgreSQLStorage/postgreshandler.py:230]: The `is_delete` column is added as the mark value to speed up snapshot and syncing. Generally, filtering on boolean would be faster than filtering on bytes.

File: jinahub/indexers/storage/PostgreSQLStorage/postgreshandler.py [hunk 13]
```diff
@@ -368,8 +420,8 @@ def get_snapshot(self, shards_to_get: List[int]):
             cursor.itersize = 10000
             cursor.execute(
                 f'SELECT doc_id, embedding from {self.snapshot_table} '
-                f'WHERE shard in %s '
-                f'ORDER BY doc_id',
+                f'WHERE shard in %s'
+                + (' and is_deleted = false' if filter_deleted else ''),
```
[the remark this answers is not in the document]
Comment (numb3r3) [reply jinahub/indexers/storage/PostgreSQLStorage/postgreshandler.py:424]: > Isn't the ordering broken now?

what do you mean?

File: jinahub/indexers/storage/PostgreSQLStorage/postgreshandler.py [hunk 14]
```diff
@@ -378,69 +430,103 @@ def get_snapshot(self, shards_to_get: List[int]):
                     if rec[1] is not None
                     else None
                 )
-                yield rec[0], vec
+                yield rec[0], vec, None, None
```
[the remark this answers is not in the document]
Comment (numb3r3) [reply jinahub/indexers/storage/PostgreSQLStorage/postgreshandler.py:433]: This change aims to make sure `get_delta_updates` and `get_snapshot` can return the same format of data.

File: jinahub/indexers/storage/PostgreSQLStorage/tests/test_postgres_dbms.py (hunk 2 above)
Comment (numb3r3) [reply jinahub/indexers/storage/PostgreSQLStorage/tests/test_postgres_dbms.py:228]: `/clear` should have the same meaning as in other indexers.

File: tests/integration/psql_import/test_import_psql.py (hunk 5 above)
Comment (numb3r3) [reply tests/integration/psql_import/test_import_psql.py:406]: `size` means the count of active documents which can be indexed in storage. Thus, the soft-deleted document should not be considered.

File: jinahub/indexers/storage/PostgreSQLStorage/postgreshandler.py (hunk 11 above)
[the remark this answers is not in the document]
Comment (numb3r3) [reply jinahub/indexers/storage/PostgreSQLStorage/postgreshandler.py:135]: The comparision results

1. this PR with using `faiss.remove_ids` for delta updating

```
benchmark@75627[I]:### indexing 100000 docs takes 1 minute and 18 seconds (78.01s)
      benchmark@75627[I]:### importing 100000 docs takes 19 seconds (19.33s)
      benchmark@75627[I]:### delete 10000 docs takes 1 second (1.53s)
      benchmark@75627[I]:### update 10000 docs takes 4 seconds (4.76s)
      benchmark@75627[I]:### delta updates 15000 docs takes 4 seconds (4.43s)
      benchmark@75627[I]:### snapshot 110000 docs takes 24 seconds (24.24s)
      benchmark@75627[I]:### importing 110000 docs takes 24 seconds (24.57s)
      benchmark@75627[I]:### searching 10 docs takes 0 seconds (0.24s)
```

2. this PR without using `faiss.remove_ids` for delta uploading
```
      benchmark@74496[I]:### indexing 100000 docs takes 1 minute and 13 seconds (73.21s)
      benchmark@74496[I]:### importing 100000 docs takes 22 seconds (22.76s)
      benchmark@74496[I]:### delete 10000 docs takes 1 second (1.55s)
      benchmark@74496[I]:### update 10000 docs takes 4 seconds (4.56s)
      benchmark@74496[I]:### delta updates 15000 docs takes 25 seconds (15.75s)
      benchmark@74496[I]:### snapshot 110000 docs takes 16 seconds (16.12s)
      benchmark@74496[I]:### importing 110000 docs takes 24 seconds (24.67s)
      benchmark@74496[I]:### searching 10 docs takes 1 second (1.48s)
```  

3. the old version
``` 
      benchmark@75859[I]:### indexing 100000 docs takes 58 seconds (58.32s)
      benchmark@75859[I]:### importing 100000 docs takes 18 seconds (18.33s)
      benchmark@75859[I]:### delete 10000 docs takes 2 seconds (2.66s)
      benchmark@75859[I]:### update 10000 docs takes 4 seconds (4.57s)
      benchmark@75859[I]:### delta updates 15000 docs takes 24 seconds (26.59s)
      benchmark@75859[I]:### snapshot 110000 docs takes 23 seconds (23.82s)
      benchmark@75859[I]:### searching 10 docs takes 2 seconds (2.84s)
```

File: jinahub/indexers/storage/PostgreSQLStorage/postgreshandler.py (hunk 13 above)
[the remark this answers is not in the document]
Comment (numb3r3) [reply jinahub/indexers/storage/PostgreSQLStorage/postgreshandler.py:424]: ah, do we need to preserve the ordering here?

File: jinahub/indexers/storage/PostgreSQLStorage/postgreshandler.py (hunk 11 above)
[the remark this answers is not in the document]
Comment (numb3r3) [reply jinahub/indexers/storage/PostgreSQLStorage/postgreshandler.py]: @cristianmtr I repeated the experiments on a Linux machine. The improvement on delta updates archived by adding indexing is not significant. Thus, I will remove the indexing in this PR.

File: tests/integration/psql_import/test_import_psql.py (hunk 4 above)
Comment (numb3r3) [reply tests/integration/psql_import/test_import_psql.py]: issue has been fixed in latest jina

**Review by cristianmtr — APPROVED** [review]

**Pull request merged**